### DISCUSSION TITLE
Best way to implement a heatmap / 2D numpy array in DuckDB?

### DISCUSSION THREAD

**Discussion by vedantroy** [duckdb/duckdb#6333] [Q&A]
I have a 300,000 x 300,000 array of float32 values that represents a similarity matrix.
Normally, I'd analyze this using numpy or pytorch, but it takes up too much space in memory (300,000^2 * 4 = 360 GB).
So, I'm trying to use duckdb instead.

Here's the schema I have so far:
```
    db.execute("CREATE TABLE pairs (idx1 INT, idx2 INT, sim FLOAT);")
```

the problem is, first inserts are extremely slow. Is there a way for me to speed them up?

Second, if I want to do queries like (find all indexes that are most similar to index 2), what would be the best way to efficiently execute these types of queries? E.g, I assume there's a few indexes I'll need.

Lastly, I'm guessing this is a mildly unconventional use-case. So, is duckdb, a good fit for a query like this?

**Comment by Mytherin** [comment]
> the problem is, first inserts are extremely slow. Is there a way for me to speed them up?

How are you inserting the data? DuckDB can directly read from many input sources (such as Pandas or Polars DataFrames, Parquet files). Inserting using those methods should be rather fast.

> Second, if I want to do queries like (find all indexes that are most similar to index 2), what would be the best way to efficiently execute these types of queries? E.g, I assume there's a few indexes I'll need.

Could you phrase that in a way we understand (i.e. SQL)? Then it would be easier to give you some recommendations :)

> Lastly, I'm guessing this is a mildly unconventional use-case. So, is duckdb, a good fit for a query like this?

We don't have any specific optimizations for matrix operations, so I doubt that it is optimal. It might be fast enough, however.

**Reply by vedantroy** [reply]
I decided to not use DUCKDB, and instead use specialized vector search tools, so I will not add any further details here for now

**Comment by l1t1** [comment]
is the matrix sparse?
what is the meaning of similar?
